core/services/relay/evm/read/event.go

```diff
@@ -0,0 +1,654 @@
+	"github.com/smartcontractkit/chainlink-common/pkg/types/query/primitives"
+
+	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/logpoller"
+	evmtypes "github.com/smartcontractkit/chainlink/v2/core/chains/evm/types"
+	"github.com/smartcontractkit/chainlink/v2/core/services/relay/evm/codec"
+	"github.com/smartcontractkit/chainlink/v2/core/services/relay/evm/types"
+)
+
+type EventBinding struct {
+	// read-only properties
+	contractName   string
+	eventName      string
+	hash           common.Hash
+	inputInfo      types.CodecEntry
+	inputModifier  commoncodec.Modifier
+	codecTopicInfo types.CodecEntry
+
+	// dependencies
+	// filterRegistrar in EventBinding is to be used as an override for lp filter defined in the contract binding.
+	// If filterRegisterer is nil, this event should be registered with the lp filter defined in the contract binding.
+	registrar      *syncedFilter
+	registerCalled bool
+	lp             logpoller.LogPoller
+
+	// internal properties / state
+	codec  commontypes.RemoteCodec
+	bound  map[common.Address]bool // bound determines if address is set to the contract binding.
+	mu     sync.RWMutex
+	topics map[string]topicDetail // topics maps a generic topic name (key) to topic data
+	// eventDataWords maps a generic name to a word index
+	// key is a predefined generic name for evm log event data word
+	// for e.g. first evm data word(32bytes) of USDC log event is value so the key can be called value
+	eventDataWords       map[string]uint8
+	confirmationsMapping map[primitives.ConfidenceLevel]evmtypes.Confirmations
+}
+
+func NewEventBinding(
+	contract, event string,
+	poller logpoller.LogPoller,
+	hash common.Hash,
+	inputInfo types.CodecEntry,
+	inputModifier commoncodec.Modifier,
+	codecTopicInfo types.CodecEntry,
+	confirmations map[primitives.ConfidenceLevel]evmtypes.Confirmations,
+) *EventBinding {
+	return &EventBinding{
+		contractName:         contract,
+		eventName:            event,
+		lp:                   poller,
+		hash:                 hash,
+		inputInfo:            inputInfo,
+		inputModifier:        inputModifier,
+		codecTopicInfo:       codecTopicInfo,
+		confirmationsMapping: confirmations,
+		topics:               make(map[string]topicDetail),
+		eventDataWords:       make(map[string]uint8),
+		bound:                make(map[common.Address]bool),
+	}
+}
+
+type topicDetail struct {
+	abi.Argument
+	Index uint64
+}
+
+var _ Reader = &EventBinding{}
+
+func (b *EventBinding) SetCodec(codec commontypes.RemoteCodec) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	b.codec = codec
+}
+
+func (b *EventBinding) Bind(ctx context.Context, bindings ...common.Address) error {
+	if b.hasBindings() {
+		// we are changing contract address reference, so we need to unregister old filter if it exists
+		if err := b.Unregister(ctx); err != nil {
+			return err
+		}
+	}
+
+	// filterRegisterer isn't required here because the event can also be polled for by the contractBinding common filter.
+	if b.registrar != nil {
```

Comment:
just add a return if registrar is nil to omit these checks

Reply:
Do you mean return early if no registrar exists?

Reply:
yup